File: Java/P003DotProduct.java
/*
Create a basic neuron layer with dot product in Java

Associated tutorial https://www.youtube.com/watch?v=tMrbN67U9d4
*/

import java.util.Arrays;

public class P003DotProduct {
  public static void main(String[] args) {
    double[] inputs = { 1.0, 2.0, 3.0, 2.5 };

    double[][] weights = {
      { 0.2, 0.8, -0.5, 1 },
      { .5, -0.91, 0.26, -0.5 },
      { -0.26, -0.27, 0.17, 0.87 }
    };

    double[] biases = { 2, 3, 0.5 };

    System.out.println(Arrays.toString(add(dotProduct(weights, inputs), biases)));
  }

  private static double[] dotProduct(double[][] input1, double[] input2) {
    double[] outputs = new double[input1.length];

    for (int i = 0; i < input1.length; i++) {
      double output = 0;
      for (int j = 0; j < input2.length; j++) {
        output += input1[i][j] * input2[j];
      }
      outputs[i] = output;
    }

    return outputs;
  }

  private static double[] add(double[] input1, double[] input2) {
    double[] output = new double[input1.length];

    for (int i = 0; i < input1.length; i++) {
      output[i] = input1[i] + input2[i];
    }

    return output;
  }
}

File: Elixir/p003_dot_product.ex
defmodule P003DotProduct do
  @moduledoc """
  Create a basic neuron layer with dot product in Elixir

  Associated tutorial https://www.youtube.com/watch?v=tMrbN67U9d4
  """

  import Enum

  def calculate() do
    inputs = [1.0, 2.0, 3.0, 2.5]
    weights = [ [ 0.2, 0.8, -0.5, 1.0],
                [0.5, -0.91, 0.26, -0.5],
                [-0.26, -0.27, 0.17, 0.87] ]
    biases = [2.0, 3.0, 0.5]
    dot(inputs, weights, biases)
    |> IO.inspect
  end

  defp dot(inputs, weight), do:
    0..length(inputs) - 1
    |> reduce(0, fn i, acc -> acc + at(inputs, i) * at(weight, i) end)
  defp dot(inputs, weights, biases), do:
    0..length(weights) - 1
    |> reduce([], fn i, acc -> acc ++ [dot(inputs, at(weights, i)) + at(biases, i)] end)
    |> List.flatten

end
